File: cpp/audio/audio_capture.cc
/**
 * @file audio_capture.cc
 * @brief Implementation of the AudioCapture class
 */

#include "audio_capture.h"
#include <portaudio.h>
#include <cmath>
#include <iostream>
#include <fstream>
#include <chrono>
#include <algorithm>

namespace koelingo {
namespace audio {

// AudioCapture constructor
AudioCapture::AudioCapture(int sample_rate, int chunk_size, int channels, int format_type)
    : sample_rate_(sample_rate),
      chunk_size_(chunk_size),
      channels_(channels),
      format_type_(format_type),
      pa_(nullptr),
      stream_(nullptr),
      is_recording_(false),
      audio_level_callback_(nullptr),
      recording_thread_(nullptr) {

    // Initialize PortAudio
    PaError err = Pa_Initialize();
    if (err != paNoError) {
        std::cerr << "PortAudio initialization error: " << Pa_GetErrorText(err) << std::endl;
        return;
    }

    pa_ = reinterpret_cast<void*>(1); // Mark as initialized

    // Calculate max buffer size (30 seconds of audio)
    max_buffer_size_ = static_cast<int>(buffer_seconds_ * sample_rate_ / chunk_size_);
}

// AudioCapture destructor
AudioCapture::~AudioCapture() {
    // Stop recording if active
    stop_recording();

    // Terminate PortAudio
    if (pa_) {
        Pa_Terminate();
        pa_ = nullptr;
    }
}

// Start recording audio
bool AudioCapture::start_recording(std::function<void(float)> audio_level_callback) {
    if (is_recording_) {
        return true;
    }

    if (!pa_) {
        std::cerr << "PortAudio not initialized" << std::endl;
        return false;
    }

    audio_level_callback_ = audio_level_callback;

    // Clear the audio buffer
    {
        std::lock_guard<std::mutex> lock(buffer_mutex_);
        audio_buffer_.clear();
    }

    // Open a PortAudio stream
    PaStreamParameters inputParams;
    inputParams.device = Pa_GetDefaultInputDevice();
    if (inputParams.device == paNoDevice) {
        std::cerr << "No default input device" << std::endl;
        return false;
    }

    inputParams.channelCount = channels_;
    inputParams.sampleFormat = format_type_;
    inputParams.suggestedLatency = Pa_GetDeviceInfo(inputParams.device)->defaultLowInputLatency;
    inputParams.hostApiSpecificStreamInfo = nullptr;

    PaError err = Pa_OpenStream(
        reinterpret_cast<PaStream**>(&stream_),
        &inputParams,
        nullptr,  // No output
        sample_rate_,
        chunk_size_,
        paClipOff,
        reinterpret_cast<PaStreamCallback*>(&AudioCapture::audio_callback),
        this
    );

    if (err != paNoError) {
        std::cerr << "Error opening PortAudio stream: " << Pa_GetErrorText(err) << std::endl;
        return false;
    }

    // Start the stream
    err = Pa_StartStream(reinterpret_cast<PaStream*>(stream_));
    if (err != paNoError) {
        std::cerr << "Error starting PortAudio stream: " << Pa_GetErrorText(err) << std::endl;
        Pa_CloseStream(reinterpret_cast<PaStream*>(stream_));
        stream_ = nullptr;
        return false;
    }

    is_recording_ = true;

    // Start background processing thread
    recording_thread_ = std::make_unique<std::thread>(&AudioCapture::process_audio, this);

    return true;
}

// Stop recording audio
void AudioCapture::stop_recording() {
    if (!is_recording_) {
        return;
    }

    is_recording_ = false;

    // Close the PortAudio stream
    if (stream_) {
        Pa_StopStream(reinterpret_cast<PaStream*>(stream_));
        Pa_CloseStream(reinterpret_cast<PaStream*>(stream_));
        stream_ = nullptr;
    }

    // Wait for the processing thread to finish
    if (recording_thread_ && recording_thread_->joinable()) {
        recording_thread_->join();
        recording_thread_.reset();
    }
}

// Get the current audio buffer
std::vector<char> AudioCapture::get_buffer() const {
    std::lock_guard<std::mutex> lock(buffer_mutex_);

    // Calculate total size
    size_t totalSize = 0;
    for (const auto& chunk : audio_buffer_) {
        totalSize += chunk.size();
    }

    // Concatenate all chunks
    std::vector<char> buffer;
    buffer.reserve(totalSize);
    for (const auto& chunk : audio_buffer_) {
        buffer.insert(buffer.end(), chunk.begin(), chunk.end());
    }

    return buffer;
}

// Save the audio buffer to a WAV file
bool AudioCapture::save_buffer_to_file(const std::string& filename) const {
    std::vector<char> buffer = get_buffer();
    if (buffer.empty()) {
        std::cerr << "Buffer is empty, nothing to save" << std::endl;
        return false;
    }

    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        std::cerr << "Failed to open file: " << filename << std::endl;
        return false;
    }

    // WAV file header
    struct WAVHeader {
        char riff[4] = {'R', 'I', 'F', 'F'};
        uint32_t fileSize = 0; // Will be set later
        char wave[4] = {'W', 'A', 'V', 'E'};
        char fmt[4] = {'f', 'm', 't', ' '};
        uint32_t fmtChunkSize = 16;
        uint16_t audioFormat = 1; // PCM
        uint16_t numChannels = 0; // Will be set
        uint32_t sampleRate = 0; // Will be set
        uint32_t byteRate = 0; // Will be set
        uint16_t blockAlign = 0; // Will be set
        uint16_t bitsPerSample = 0; // Will be set
        char data[4] = {'d', 'a', 't', 'a'};
        uint32_t dataSize = 0; // Will be set later
    } header;

    // Set header values
    header.numChannels = static_cast<uint16_t>(channels_);
    header.sampleRate = static_cast<uint32_t>(sample_rate_);

    // Assuming format_type is paInt16, which is 16-bit PCM
    header.bitsPerSample = 16;
    header.blockAlign = header.numChannels * (header.bitsPerSample / 8);
    header.byteRate = header.sampleRate * header.blockAlign;

    header.dataSize = static_cast<uint32_t>(buffer.size());
    header.fileSize = 36 + header.dataSize;

    // Write header
    file.write(reinterpret_cast<const char*>(&header), sizeof(header));

    // Write audio data
    file.write(buffer.data(), buffer.size());

    return file.good();
}

// Calculate audio level from raw data
float AudioCapture::calculate_audio_level(const std::vector<char>& audio_data) const {
    if (audio_data.empty()) {
        return 0.0f;
    }

    // Convert bytes to 16-bit samples
    const int16_t* samples = reinterpret_cast<const int16_t*>(audio_data.data());
    size_t sample_count = audio_data.size() / sizeof(int16_t);

    // Calculate RMS
    float sum = 0.0f;
    for (size_t i = 0; i < sample_count; i++) {
        float sample = samples[i] / 32768.0f; // Normalize to [-1.0, 1.0]
        sum += sample * sample;
    }

    float rms = sqrtf(sum / sample_count);

    // Convert to dB
    float db = 20.0f * log10f(std::max(rms, 0.0000001f)); // Avoid log(0)

    // Normalize to [0.0, 1.0] range, assuming -60dB is silence
    float normalized = std::max(0.0f, (db + 60.0f) / 60.0f);
    return std::min(normalized, 1.0f);
}

// Audio processing thread
void AudioCapture::process_audio() {
    while (is_recording_) {
        // Sleep to avoid busy waiting
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
}

// Static callback for PortAudio
int AudioCapture::audio_callback(const void* input_buffer, void* output_buffer [[maybe_unused]],
                              unsigned long frames_per_buffer,
                              const PaStreamCallbackTimeInfo* time_info [[maybe_unused]], PaStreamCallbackFlags status_flags [[maybe_unused]],
                              void* user_data) {
    AudioCapture* self = static_cast<AudioCapture*>(user_data);

    if (!input_buffer || !self) {
        return paContinue;
    }

    // Calculate buffer size in bytes
    size_t buffer_size = frames_per_buffer * self->channels_ * (self->format_type_ == paInt16 ? 2 : 4);

    // Create a new buffer for this chunk
    std::vector<char> buffer(static_cast<const char*>(input_buffer),
                           static_cast<const char*>(input_buffer) + buffer_size);

    // Calculate audio level if callback is set
    if (self->audio_level_callback_) {
        float level = self->calculate_audio_level(buffer);
        self->audio_level_callback_(level);
    }

    // Add buffer to the queue
    {
        std::lock_guard<std::mutex> lock(self->buffer_mutex_);
        self->audio_buffer_.push_back(std::move(buffer));

        // Limit buffer size
        while (static_cast<int>(self->audio_buffer_.size()) > self->max_buffer_size_) {
            self->audio_buffer_.pop_front();
        }
    }

    return paContinue;
}

// Get available audio devices
std::vector<std::map<std::string, std::variant<int, std::string>>> AudioCapture::get_available_devices() const {
    std::vector<std::map<std::string, std::variant<int, std::string>>> devices;

    if (!pa_) {
        return devices;
    }

    int numDevices = Pa_GetDeviceCount();
    for (int i = 0; i < numDevices; i++) {
        const PaDeviceInfo* deviceInfo = Pa_GetDeviceInfo(i);
        if (deviceInfo && deviceInfo->maxInputChannels > 0) {
            std::map<std::string, std::variant<int, std::string>> device;
            device["index"] = i;
            device["name"] = std::string(deviceInfo->name);
            device["channels"] = deviceInfo->maxInputChannels;
            devices.push_back(device);
        }
    }

    return devices;
}

} // namespace audio
} // namespace koelingo

File: src/audio/audio_capture.py
"""
Audio capture module for KoeLingo.

This module handles microphone input and audio processing.
"""

import pyaudio
import numpy as np
import wave
import threading
import time
from typing import Optional, Callable, Tuple
from collections import deque


class AudioCapture:
    """Audio capture and processing class for real-time audio input."""

    def __init__(
        self,
        sample_rate: int = 16000,
        chunk_size: int = 1024,
        channels: int = 1,
        format_type: int = pyaudio.paInt16,
    ):
        """
        Initialize the audio capture module.

        Args:
            sample_rate: Sample rate in Hz (default: 16kHz for Whisper)
            chunk_size: Number of frames per buffer
            channels: Number of audio channels (1 for mono, 2 for stereo)
            format_type: Audio format type from pyaudio
        """
        self.sample_rate = sample_rate
        self.chunk_size = chunk_size
        self.channels = channels
        self.format_type = format_type

        self.audio = pyaudio.PyAudio()
        self.stream = None
        self.is_recording = False
        self.audio_level_callback = None
        self._recording_thread = None

        # Audio buffer for storing chunks (30 seconds of audio)
        self.buffer_seconds = 30
        self.max_buffer_size = int(self.buffer_seconds * self.sample_rate / self.chunk_size)
        self.audio_buffer = []

        # Frame counter for testing purposes
        self.frame_count = 0

        # Audio queue for processing
        self.audio_queue = deque(maxlen=100)

        # Selected device index
        self.selected_device_index = None

    def start_recording(self, audio_level_callback: Optional[Callable[[float], None]] = None) -> bool:
        """
        Start recording audio from the microphone.

        Args:
            audio_level_callback: Optional callback function to receive audio level updates

        Returns:
            bool: True if recording started successfully, False otherwise
        """
        if self.is_recording:
            return True

        try:
            self.audio_level_callback = audio_level_callback

            # Use selected device if specified
            input_device = None
            if self.selected_device_index is not None:
                input_device = self.selected_device_index

            self.stream = self.audio.open(
                format=self.format_type,
                channels=self.channels,
                rate=self.sample_rate,
                input=True,
                input_device_index=input_device,
                frames_per_buffer=self.chunk_size,
                stream_callback=self._audio_callback
            )

            self.is_recording = True
            self.audio_buffer = []
            self.frame_count = 0
            self.audio_queue.clear()

            # Start a thread to process audio in background
            self._recording_thread = threading.Thread(target=self._process_audio)
            self._recording_thread.daemon = True
            self._recording_thread.start()

            return True
        except Exception as e:
            print(f"Error starting audio recording: {e}")
            return False

    def stop_recording(self) -> None:
        """Stop recording audio from the microphone."""
        if not self.is_recording:
            return

        self.is_recording = False
        if self.stream:
            self.stream.stop_stream()
            self.stream.close()
            self.stream = None

        if self._recording_thread and self._recording_thread.is_alive():
            self._recording_thread.join(timeout=1.0)

    def _audio_callback(self, in_data, frame_count, time_info, status):
        """Callback function for audio stream."""
        if self.is_recording:
            # Add the chunk to our buffer
            self.audio_buffer.append(in_data)
            self.frame_count += 1

            # Add to processing queue
            self.audio_queue.append(in_data)

            # Keep buffer at maximum size
            while len(self.audio_buffer) > self.max_buffer_size:
                self.audio_buffer.pop(0)

            return (in_data, pyaudio.paContinue)
        return (in_data, pyaudio.paComplete)

    def _process_audio(self) -> None:
        """Process audio in a background thread."""
        while self.is_recording:
            # If we have data and a callback for audio levels
            if self.audio_buffer and self.audio_level_callback:
                # Calculate audio level from the latest chunk
                latest_chunk = self.audio_buffer[-1]
                audio_array = np.frombuffer(latest_chunk, dtype=np.int16)
                audio_level = self._calculate_audio_level(audio_array)

                # Call the callback with the audio level
                self.audio_level_callback(audio_level)

            # Sleep to avoid using too much CPU
            time.sleep(0.05)

    def _calculate_audio_level(self, audio_array: np.ndarray) -> float:
        """
        Calculate the audio level from a numpy array of audio samples.

        Args:
            audio_array: Numpy array of audio samples

        Returns:
            float: Audio level between 0.0 and 1.0
        """
        # Calculate RMS (root mean square) as audio level
        if len(audio_array) == 0:
            return 0.0

        # Normalize to range 0.0 - 1.0
        rms = np.sqrt(np.mean(np.square(audio_array.astype(np.float32))))
        normalized = min(1.0, rms / 32768.0)  # 16-bit max value is 32768

        return normalized

    def get_buffer(self) -> bytes:
        """
        Get the current audio buffer.

        Returns:
            bytes: Audio data from the buffer
        """
        return b''.join(self.audio_buffer)

    def get_buffer_as_numpy(self) -> np.ndarray:
        """
        Get the current audio buffer as a numpy array.

        Returns:
            np.ndarray: Audio data as a numpy array
        """
        buffer_bytes = self.get_buffer()
        return np.frombuffer(buffer_bytes, dtype=np.int16)

    def save_buffer_to_file(self, filename: str) -> bool:
        """
        Save the current audio buffer to a WAV file.

        Args:
            filename: Name of the file to save

        Returns:
            bool: True if saved successfully, False otherwise
        """
        try:
            buffer_data = self.get_buffer()
            with wave.open(filename, 'wb') as wf:
                wf.setnchannels(self.channels)
                wf.setsampwidth(self.audio.get_sample_size(self.format_type))
                wf.setframerate(self.sample_rate)
                wf.writeframes(buffer_data)
            return True
        except Exception as e:
            print(f"Error saving audio to file: {e}")
            return False

    def get_available_devices(self) -> list:
        """
        Get a list of available audio input devices.

        Returns:
            list: List of dictionaries with device information
        """
        devices = []
        for i in range(self.audio.get_device_count()):
            device_info = self.audio.get_device_info_by_index(i)
            # Only include input devices
            if device_info["maxInputChannels"] > 0:
                devices.append({
                    "index": i,
                    "name": device_info["name"],
                    "channels": device_info["maxInputChannels"],
                    "sample_rate": int(device_info["defaultSampleRate"])
                })
        return devices

    def select_device(self, device_index: int) -> bool:
        """
        Select an input device for recording.

        Args:
            device_index: Index of the device to use

        Returns:
            bool: True if device was selected successfully, False otherwise
        """
        try:
            # Verify the device index is valid
            self.audio.get_device_info_by_index(device_index)
            self.selected_device_index = device_index
            return True
        except Exception:
            return False

    def get_frame_count(self) -> int:
        """
        Get the number of audio frames recorded.

        Returns:
            int: Number of audio frames
        """
        return self.frame_count

    def get_queue_size(self) -> int:
        """
        Get the current size of the audio queue.

        Returns:
            int: Number of chunks in the queue
        """
        return len(self.audio_queue)

    def get_buffer_size(self) -> int:
        """
        Get the size of audio buffer in samples.

        Returns:
            int: Buffer size in samples
        """
        return self.chunk_size

    def has_signal(self) -> bool:
        """
        Check if there is a meaningful audio signal present.

        Returns:
            bool: True if signal is detected, False otherwise
        """
        if not self.audio_buffer:
            return False

        # Convert last buffer to numpy array
        audio_array = np.frombuffer(self.audio_buffer[-1], dtype=np.int16)

        # Calculate RMS level
        level = self._calculate_audio_level(audio_array)

        # Return True if level is above threshold
        return level > 0.05  # Arbitrary threshold

    def __del__(self):
        """Clean up resources when object is deleted."""
        self.stop_recording()
        if self.audio:
            self.audio.terminate()


# Example callback function
def print_audio_level(level):
    """Print the audio level."""
    bars = int(level * 50)
    display = '#' * bars + ' ' * (50 - bars)
    print(f"\rLevel: [{display}] {level:.2f}", end='')


if __name__ == "__main__":
    # Example usage
    audio_capture = AudioCapture()
    print("Available devices:")
    devices = audio_capture.get_available_devices()
    for i, device in enumerate(devices):
        print(f"{i}: {device['name']}")

    try:
        audio_capture.start_recording(print_audio_level)
        print("Recording... Press Ctrl+C to stop")
        while True:
            time.sleep(0.1)
    except KeyboardInterrupt:
        print("\nStopping recording...")
    finally:
        audio_capture.stop_recording()
        audio_capture.save_buffer_to_file("recorded_audio.wav")
        print("\nSaved audio to 'recorded_audio.wav'")